### PR TITLE
[SYCL][PI][L0] Add blitter engine support in level zero plugin

### PR DESCRIPTION
**Pull request by asudarsa**
Level Zero plugin submits all operation to level zero compute queues and both compute and memory operations are submitted to be executed on the compute engine. This change submits some of the memory operations to a level zero copy queue and such operations will be executed on the copy engine.

Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (kbsmith-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4616]: Hmm, I wonder if a memory fill will really benefit from using a copy engine.

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:610]: We access ZeCommandListFenceMap in other locations as well without a lock. We can add a lock as part of a follow-up checkin. If acceptable, I can add a TODO.

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:4616]: Specification says copy engine can handle fill operations.

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (kbsmith-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:610]: It seems a very small change so I think it should probably be addressed now. I think we should very quickly fix other places where we access without locking.

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (kbsmith-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:4616]: My comment wasn't about correctness, but whether it made sense from a performance point of view.

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (kbsmith-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:610]: I was wrong.  There is no need to further guard this.  It is already guarded by a lock on the queue mutex which the caller must already hold, and in one user, that lock is already acquired.

**Review by kbobrovs — COMMENTED** [review]
It would be good to add an end-to-end test for the new feature.

**Comment by asudarsa** [comment]
I will add a test to the llvm-test-suite. Thanks

**Comment by asudarsa** [comment]
Testing has completed. Test restarted after I updated the documentation.

File: sycl/doc/EnvironmentVariables.md
Comment (smaslov-intel) [inline sycl/doc/EnvironmentVariables.md:33]: 1) and also for "fill" operations
2) I think we should explicitly that this doesn't force blitter, but just allows. final decision is by heuristics. and then maybe we should add force semantics too?

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:366]: I suggest naming it PreferCopyEngine

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:520]: I think it is even more important to note when it was requested but not found

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:152]: add a small comment plz

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:230]: update the comment to say that it serves both caches of command-lists, but we could split this into two if needed

Comment (kbsmith-intel) [reply sycl/plugins/level_zero/pi_level_zero.hpp:230]: Please don't split it.  I was the one that requested it only be one.  Ideally (but certainly not for thsi changeset) the whole command list caching scheme itself would be hidden behind a programmatic interface, and that you'd just request a command list of the proper type, and not be worried how the caches were implemented under the covers.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.hpp:230]: I am fine with having just one as I don't believe it could be a perf. bottleneck.
But please indicate this in the comment

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:311]: describe them separately, and mention that the copy-queue might not exists at all

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:365]: add comment

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:807]: This appears very ineffective searching the entire cache of command-lists on every new submission. Can't you just grab this info directly from the fence created for this command list and passed you in `ZeFence` argument.

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:375]: I think you don't need this, see my comment at where it is used

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:150]: add a comment about what -1 value means

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:621]: name PreferCopyCommandList

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:625]: Add TODO that we'd want to add batching for copy-commands as well

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:311]: I believe that dividing the commands stream into 2 queues creates serious synchronization issues for in-order queue. Now the plugin needs to do something more to ensure that commands are executed one after another, and this something is crossing a Level-Zero queue boundary.

Tag @rbegam for awareness

Comment (kbsmith-intel) [reply sycl/plugins/level_zero/pi_level_zero.hpp:311]: I think rather than relying on the underlying run-times for ensuring in-order queues, that we should do one of two different things.
1 - We could just have the DPC++ run-time create an dependence of on each submission with the previous submission.  This is easily implemnted by having the RT place a wait on event from the previous submission on the current submission.
2 - The level zero plugin can, for each queue, have a field in the queue that is the event for the previously created event, and it can implicitly add that to the wait on event list when doing a submission.
I think either of these methods is probably better than relying on the underlying level zero runtime queues inorder properties.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.hpp:311]: Yes, that's what I meant, we need to add [many] more dependencies for these 2 queues to truly implement in-order semantics, when required. I do think we have to continue rely in what are the underlying runtimes guarantee, since that it is going to remove the redundant resource/performance overhead. For example, OpenCL RT does support in-order queues internally, so we don't want to do (1). Level-Zero RT guarantees in-order execution of commands within the same command-list, and we'd use that when implement something like you describe in (2) to minimize on events.

I think that we have to address this issue in this PR since this is breaking current in-order queues behavior. I'd start by just disabling use of a blitter-engines for in-order SYCL queues, and add proper sync separately.

File: sycl/doc/EnvironmentVariables.md
Comment (asudarsa) [reply sycl/doc/EnvironmentVariables.md:33]: I modified the comment to include 'fill' operations. Since the heuristics controlling the blitter usage are beyond the plugin scope, I am not sure if we need to add anything here.

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:366]: Agreed. I have changed it in all sites.

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:621]: I used IsCopyCommand because this boolean value specifies if incoming command is a copy command or not. Hope that sounds reasonable.

Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:621]: Upon second thoughts..I agree with name change. Thanks

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:625]: Done

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:807]: To get that, we will need a level zero API call which might be expensive. One of the ways to make this faster will be to replace std::map with std::unordered_map. Is that a viable option?

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:4616]: Added a comment about perf improvements by using blitter

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:150]: Done

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:152]: Done

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:251]: Done

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:311]: Done

Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:311]: We have a resolution which will need to be in sync with the overall effort to support in-order queues in level zero. Under discussion.

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:365]: Done

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:375]: I think we need this in order to avoid a ZE API call.

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:230]: Added comment

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: I added a comment about my performance analysis runs for memory fill operation

File: sycl/doc/EnvironmentVariables.md
Comment (asudarsa) [reply sycl/doc/EnvironmentVariables.md]: Modified comment according to suggestion. Will add an E2E test to make sure copy engine is being found by plugin when expected.

File: sycl/doc/EnvironmentVariables.md
Comment (asudarsa) [reply sycl/doc/EnvironmentVariables.md]: Modified text. Since heuristics is not part of plugin code, I suggest we do not add any 'force' semantics here.

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Agreed. Done

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Agreed. Done

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:155]: Done

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:234]: Added a comment. Thanks

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp]: Agreed. Done. Thanks

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:317]: Done. Thanks

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:152]: Done. Thanks

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Done. Thanks

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Done. Thanks

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:317]: We have a resolution that will temporarily disable blitter engine use for in-order queues. Need some discussion with owner of in-order queue supporting effort.

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:372]: Done. Thanks

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:812]: It seems like 'grabbing' this info from the fence created will require a ZE API call. I wanted to avoid that. To make this faster, we can use std::unordered_map (O(1) access) instead of std::map (O(log(n))) access time. Please advise if that is acceptable.

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:382]: It seems like 'grabbing' this info from the fence created will require a ZE API call. I wanted to avoid that. To make this faster, we can use std::unordered_map (O(1) access) instead of std::map (O(log(n))) access time. Please advise if that is acceptable.

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (kbsmith-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:812]: I don't think there is any ZE API call that can get this information.

I think the suggestion was to eliminate the query entirely by passing the inforrmation about what "kind" of command list you have back from getAvailableCommandList, and then to pass that directly into executeCommandList, and then this would just be a reference to that parameter.  Then there would be no need to query whether a command list was a copy or compute command at all.
If you do that, then when you desire to implement batching for copy engine, then in addition to the ZeOpenCommandList that is in the _pi_queue data structure, you will also need a member that indicates whether ZeOpenCommandList is a copy or compute command list, and that will be used whenever you need to call executeOpenCommandList to pass which kind of command list there is.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:812]: > It seems like 'grabbing' this info from the fence created will require a ZE API call.

No, why? You'd just return the `_pi_queue::command_list_fence_t` instead of the `ze_fence_handle_t` itself in the `getAvailableCommandList`, and take the info about the queue from it in the `executeCommandList`

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:812]: Please add TODO that this is going to be restructured, and the map lookup gone.

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4292]: I think you should check for the copy-queue to be non-null here

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4395]: Why do you negate the "PreferCopyEngine" here and in multiple other places too?

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4568]: This is not self-explanatory, put a comment what/why this check

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5629]: Add a comment explaining why you pass "false" here, and add a TODO how this could be improved.

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.hpp:317]: Are you going to add disabling for in-order queue?

**Review by smaslov-intel — CHANGES_REQUESTED** [review]
Please address more comments

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:812]: I added a TODO before the function definition.

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:4395]: AH..This is a logical error. I was replacing IsDeviceLocalCopy which would mean 'do not prefer copy engine'. I will correct this. Thanks for catching this. 
Also, currently, I map copy operations to 'compute' engines when we do not have enough data to say if it is a device-device copy. DO we want to keep it this way or do we want  to map them to 'copy' engines?

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:4568]: Done.

File: sycl/plugins/level_zero/pi_level_zero.cpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5629]: Done. I also replaced uses of 'PreferCopyCommandList' and replaced it with 'PreferCopyEngine' to be less confusing.

File: sycl/plugins/level_zero/pi_level_zero.hpp
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:317]: Added, Please see Line 572 and 653.

**Review by kbobrovs — DISMISSED** [review]
LGTM, given that E2E test will appear in llvm-test-suite

**Review by smaslov-intel — APPROVED** [review]
LGTM

**Review by kbsmith-intel — DISMISSED** [review]
LGTM

**Comment by asudarsa** [comment]
I fixed  a merge issue and recommitted....

**Review by smaslov-intel — APPROVED** [review]
After this is landed, please do rework the need of a `getZeCommandListIsCopyList` to avoid searching the map, as we discussed offline.

**Review by kbsmith-intel — APPROVED** [review]
LGTM

**Review by kbobrovs — APPROVED** [review]

**Pull request merged**
